File: 2019/src/main/scala/Day8.scala
import scala.collection.mutable.ArrayBuffer

object Day8 {
  val input = scala.io.Source.fromFile("data/data8.txt", "UTF-8").getLines.toVector

  def part1(): Unit = {
    var list = input(0).toList
    var resTot:ArrayBuffer[ArrayBuffer[Int]] = new ArrayBuffer()
    object Done extends Exception{}
    var i = 0
    try {
      while (true){
        var res:ArrayBuffer[Int] = new ArrayBuffer()
        for (k <- 0 until 25){
          for (l <- 0 until 6){
            if (k*6+l + i*25*6 >= list.length) throw Done
            res += list(l+k*6 + i*25*6).toInt - 48
          }
        }
        resTot += res
        i+=1
      }
    }
    catch {
      case Done =>
    }
    val counting = resTot.map(xs => xs.count(_==0))
    val index = counting.indexOf(counting.min)
    val ones = resTot(index).count(_ == 1)
    val twos = resTot(index).count(_ == 2)
    println(resTot(index))
    println("ones: " + ones + " * twos: " + twos + " = " + ones*twos)
  }

  def part2(): Unit = {
    var layers = input(0).grouped(25*6).toVector

    var image:String = ""

    for (i <- layers(0).indices) {
      var quit = false
      var k = 0
      while(!quit) {
        if (layers(k)(i) == '0') {
          image += ' '
          quit = true
        } else if (layers(k)(i) == '1') {
          image += '#'
          quit = true
        }
        k+=1
      }
      if ((i+1) % 25 == 0) image += '\n'
    }
    
    println(image)
  }

  def apply() = {
    println("SOLUTION DAY 8")
    println("Running part 1")
    //part1()
    println("Running part 2")
    part2()
  }
}

File: 2019/python/src/day8.py
import os
import sys

def main():
    input = []
    
    with open(f'{os.path.dirname(os.path.realpath(__file__))}/../input/day8.txt') as f:
        for line in f.readlines():
            input.append(line.replace("\n",""))

    print("Part 1: " + str(part1(input)))
    print("Part 2: " + str(part2(input)))

def part1(input):
    data = [input[0][i:i+25] for i in range(0, len(input[0]), 25)]
    layers = [data[i:i+6] for i in range(0, len(data), 6)]

    minimum = (10000000, 0, 0)
    
    for layer in layers:
        zeros = 0
        ones = 0
        twos = 0
        for row in layer:
            for pixel in row:
                if pixel == '0':
                    zeros += 1
                if pixel == '1':
                    ones += 1
                if pixel == '2':
                    twos += 1

        if zeros < minimum[0]:
            minimum = (zeros, ones, twos)
    
    return minimum[1] * minimum[2]

def part2(input):
    data = [input[0][i:i+25] for i in range(0, len(input[0]), 25)]
    layers = [data[i:i+6] for i in range(0, len(data), 6)]

    image = [['2']*25 for _ in range(6)]
    
    for layer in layers:
        for r in range(len(layer)):
            for c in range(len(layer[r])):
                if image[r][c] == '2':
                    image[r][c] = layer[r][c]
    
    return "\n"+"\n".join(["".join(row).replace('0', ' ').replace('1', '#') for row in image])

if __name__ == "__main__":
    sys.exit(main())